smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java

```diff
@@ -0,0 +1,376 @@
+     */
+    public static TraceFile fromNode(Node value) {
+        ObjectNode node = value.expectObjectNode();
+        Builder builder = builder()
+                .smithyTrace(node.expectStringMember(SMITHY_TRACE_TEXT).getValue())
+                .artifact(ArtifactMetadata.fromNode(node.expectObjectMember(ARTIFACT_TEXT)));
+
+        //parse shapes
+        Map<StringNode, Node> shapeMap = node.expectObjectMember(SHAPES_TEXT).getMembers();
+        for (Map.Entry<StringNode, Node> entry : shapeMap.entrySet()) {
+            for (Node linkNode : (entry.getValue().expectArrayNode()).getElements()) {
+                builder.addShapeLink(entry.getKey().getValue(), ShapeLink.fromNode(linkNode));
+            }
+        }
+
+        //parse definitions
+        if (node.containsMember(DEFINITIONS_TEXT)) {
+            builder.definitions(ArtifactDefinitions.fromNode(node.expectObjectMember(DEFINITIONS_TEXT)));
+        }
+
+        return builder.build();
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Converts TraceFile instance variables into an
+     * ObjectNode.
+     *
+     * @return ObjectNode representation of a TraceFile.
+     */
+    @Override
+    public ObjectNode toNode() {
+        //constructing shapes ObjectNode map
+        ObjectNode.Builder shapesBuilder = ObjectNode.objectNodeBuilder();
+        for (Map.Entry<ShapeId, List<ShapeLink>> entry : shapes.entrySet()) {
+            String shapeId = entry.getKey().toString();
+            ArrayNode shapeListNode = entry.getValue() //get list of ShapeLinks
+                    .stream()
+                    .map(ShapeLink::toNode) //convert each ShapeLink to an ObjectNode
+                    .collect(ArrayNode.collect()); //collect each ObjectNode in an ArrayNode
+            shapesBuilder.withMember(shapeId, shapeListNode);
+        }
+
+        //returning ObjectNode for TraceFile
+        return ObjectNode.objectNodeBuilder()
+                .withMember(SMITHY_TRACE_TEXT, smithyTrace)
+                .withMember(ARTIFACT_TEXT, artifactMetadata)
+                .withOptionalMember(DEFINITIONS_TEXT, getArtifactDefinitions())
+                .withMember(SHAPES_TEXT, shapesBuilder.build())
+                .build();
+    }
+
+    /**
+     * Throws an error if any ShapeLink object contains a tag or type that is not in artifactDefinition's.
+     * This method should be called after creating a TraceFile object to verify that all the types and tags
+     * in shapes have been defined in artifactDefinition's. This TraceFile's ArtifactDefinitions object
+     * MUST be defined prior to calling this method.
+     *
+     * @throws ExpectationNotMetException if a type or tag in shapes is not in artifactDefinitions, or if
+     *                                    artifactDefinitions is not defined when the method is called.
+     */
+    public void validateTypesAndTags() {
```

Comment:
Should this validation be `private` and automatically run after setting the properties on this object in the constructor? Otherwise, it's possible for an invalid `TraceFile` instance to be used.

```java
private TraceFile(Builder builder) {
    smithyTrace = SmithyBuilder.requiredState(SMITHY_TRACE_TEXT, builder.smithyTrace);
    artifactMetadata = SmithyBuilder.requiredState(ARTIFACT_TEXT, builder.artifactMetadata);
    if (builder.shapes.isEmpty()) {
        throw new IllegalStateException("TraceFile's shapes field must not be empty to build it.");
    }
    shapes = MapUtils.copyOf(builder.shapes);
    artifactDefinitions = builder.artifactDefinitions;

    // Validate we received a well-formed TraceFile
    validateTypesAndTags();
}
```